opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/servlet/TracingHandlerInterceptor.java

```diff
@@ -0,0 +1,103 @@
+import io.opentracing.tag.Tags;
+
+/**
+ * Tracing handler interceptor for spring web. It creates a new span for an incoming request.
+ * Created span can be accessed via {@link HttpServletRequest#getAttribute(String)} with name
+ * {@link TracingHandlerInterceptor#ACTIVE_SPAN}.
+ *
+ * @author Pavol Loffay
+ */
+public class TracingHandlerInterceptor extends HandlerInterceptorAdapter {
+
+    /**
+     * Used as a name of {@link HttpServletRequest#setAttribute(String, Object)} to inject current server span
+     */
+    public static final String ACTIVE_SPAN = TracingHandlerInterceptor.class.getName() + ".activeSpan";
+    private static final String ACTIVE_SPAN_WRAPPER = TracingHandlerInterceptor.class.getName() + ".activeSpan";
+
+    private Tracer tracer;
+    private List<SpanDecorator> decorators;
+    private TracingDecision tracingDecision;
+
+    /**
+     * @param tracer tracer
+     * @param decorators span decorators
+     * @param tracingDecision tracing decision
+     */
+    public TracingHandlerInterceptor(Tracer tracer, List<SpanDecorator> decorators, TracingDecision tracingDecision) {
+        this.tracer = tracer;
+        this.decorators = new ArrayList<>(decorators);
+        this.tracingDecision = tracingDecision;
+    }
+
+    @Override
+    public boolean preHandle(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse, Object handler)
+            throws Exception {
+
+        Span serverSpan = (Span) httpServletRequest.getAttribute(TracingFilter.ACTIVE_SPAN);
+        SpanWrapper spanWrapper = (SpanWrapper) httpServletRequest.getAttribute(ACTIVE_SPAN_WRAPPER);
+
+        if (serverSpan != null && spanWrapper == null) {
+            for (SpanDecorator decorator: decorators) {
+                decorator.onPreHandle(httpServletRequest, handler, serverSpan);
+            }
+            spanWrapper = new SpanWrapper(serverSpan, true);
+        } else if (httpServletRequest.getAttribute(ACTIVE_SPAN) == null &&
+                tracingDecision.isTraced(httpServletRequest, httpServletResponse, handler)) {
+
+            SpanContext parentContext = tracer.extract(Format.Builtin.HTTP_HEADERS,
+                    new HttpServletHeadersExtractAdapter(httpServletRequest));
+
+            serverSpan = tracer.buildSpan(httpServletRequest.getMethod())
+                    .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_SERVER)
+                    .asChildOf(parentContext)
+                    .start();
+
+            for (SpanDecorator decorator: decorators) {
+                decorator.onPreHandle(httpServletRequest, handler, serverSpan);
+            }
+
+            // when exception is thrown in controller this is method is called twice, therefore using
+            // spanWrapper (span#finish()) is not idempotent by specification.
+            spanWrapper = new SpanWrapper(serverSpan, false);
+            httpServletRequest.setAttribute(ACTIVE_SPAN, spanWrapper);
```

Comment:
Is this correct? You set the `ACTIVE_SPAN` as the wrapper, but you seem to expect the `ACTIVE_SPAN` to be a `Span` (line 53)? Didn't you mean `ACTIVE_SPAN_WRAPPER`?